File: mark-track/backend/models/teacher.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

class TeacherDetails(BaseModel):
    uid: str
    first_name: str
    last_name: str
    father_name: str
    gov_number: str
    subject_id: str

class TeacherRequest(BaseModel):
    teacher_id: str

class MarkBase(BaseModel):
    student_id: str
    value: float = Field(..., ge=1, le=10)  
    subject_id: str
    date: datetime = Field(default_factory=datetime.now)

class MarkCreate(MarkBase):
    pass

class Mark(MarkBase):
    id: str
    teacher_id: str

class AbsenceBase(BaseModel):
    student_id: str
    subject_id: str
    date: datetime = Field(default_factory=datetime.now)
    is_motivated: bool = False

class AbsenceCreate(AbsenceBase):
    pass

class Absence(AbsenceBase):
    id: str
    teacher_id: str

class StudentStats(BaseModel):
    student_id: str
    first_name: str
    last_name: str
    marks: List[Mark] = []
    absences: List[Absence] = []
    average_mark: float = 0
    total_absences: int = 0
    motivated_absences: int = 0

class ClassStats(BaseModel):
    class_id: str
    name: str
    average_mark: float = 0
    total_absences: int = 0
    motivated_absences: int = 0
    students: List[StudentStats] = []

class TeacherClass(BaseModel):
    id: str
    name: str
    subject_id: str

class TeacherClasses(BaseModel):
    classes: List[TeacherClass]

class StudentResponse(BaseModel):
    id: str
    student_id: str
    first_name: str
    last_name: str
    marks: Optional[List[Mark]] = None
    absences: Optional[List[Absence]] = None
    average_mark: Optional[float] = None
    total_absences: Optional[int] = None
    motivated_absences: Optional[int] = None

class StudentsResponse(BaseModel):
    students: List[StudentResponse]

File: mark-track/src/types/teacher.ts
export interface TeacherClass {
    id: string;
    name: string;
    subject_id: string;
}

export interface TeacherClasses {
    classes: TeacherClass[];
}

export interface MarkBase {
    student_id: string;
    value: number;
    subject_id: string;
    date: string;
}

export interface MarkCreate extends MarkBase {}

export interface Mark extends MarkBase {
    id: string;
    teacher_id: string;
    description: string;
    value: number;


}

export interface AbsenceBase {
    student_id: string;
    subject_id: string;
    date: string;
    is_motivated: boolean;
}

export interface AbsenceCreate extends AbsenceBase {}

export interface Absence extends AbsenceBase {
    id: string;
    teacher_id: string;
    description: string;
    is_motivated: boolean;

}

export interface StudentStats {
    student_id: string;
    first_name: string;
    last_name: string;
    marks: Mark[];
    absences: Absence[];
    average_mark: number;
    total_absences: number;
    motivated_absences: number;
}

export interface ClassStats {
    class_id: string;
    name: string;
    average_mark: number;
    total_absences: number;
    motivated_absences: number;
    students: StudentStats[];
}

export interface StudentResponse {
    id: string;
    student_id: string;
    first_name: string;
    last_name: string;
    marks?: Mark[];
    absences?: Absence[];
    average_mark?: number;
    total_absences?: number;
    motivated_absences?: number;
}

export interface StudentsResponse {
    students: StudentResponse[];
} 